sections/testing/unit-testing.tex

```diff
@@ -0,0 +1,97 @@
+One of the first decisions when creating unit tests, is the choice of test runner.
+Since the React library is shipped with Jest \cite{ReactRunningTests}, and since both parts of the system are JavaScript based, it seemed like the optimal solution to use the same test runner.
+However, after using Jest for testing the API for a while, several issues surfaced when trying to implement mock functions related to our ORM and some TypeScript functionality in general.
+To simplify the creation of unit testing on the API, the test driver for this was changed to Mocha.
+
+One of the main differences between Jest and Mocha is that Jest is a complete suit that supports running the tests, asserting results, stubbing, test coverage and much more.
+Mocha, on the other hand, is just a test runner and requires additional libraries for asserting, stubbing and generating code coverage results \cite{Mocha}.
+We chose to use Chai for assertions, Sinon for stubbing and Nyc for generating code coverage.
+
+\subsection{What to test}
+The answer to what should be tested varies a lot between the frontend and the API.
+One of the axioms of testing states that it is impossible to test a program completely, and should be seen as a risk-based exercise where the tester asserts which parts of the system should be tested, and how thoroughly it should be tested \cite{SoftwareTesting}.
+For the frontend of the system, it would be possible to test whether each component is rendered to the user, if all the information is correct, what happens if you use 0 as input to the function, 1 as input, 3 as input.. And so on.
+Instead, we chose to focus on testing if the components that are conditionally rendered are shown when the conditions are fulfilled.
+An example of this could be the \textit{Show User} page, where the system should show tutor info if and only if the user is actually a tutor.
+This now leads to test cases like is the info rendered if:
+\begin{itemize}
+    \item the user is a tutor?
+    \item the user is a tutor?
+    \item the user is a tutor, but the tutor information is invalid?
+    \item the user is a tutor, but the tutor information is incomplete?
+\end{itemize}
+
+In addition to the conditional rendering, any business logic located on the frontend should be tested to ensure both that it works as intended, but also to ensure that any future updates to the code will not break it, which is referred to as regression testing.
+
+For the API, some functionality like the ORM is handled by external packages. Since this is not code we're writing ourselves, we have chosen to assume that the developers of the packages have tested the software and we will not have to.
+We focus on unit testing our controllers and middleware to ensure that they work as expected, and instead do some integration testing for the external packages, to ensure that they work as expected when integrated into our system. This will be further described in \autoref{sec:integrationTesting}.
```

Comment:
```suggestion
We focus on unit testing our controllers and middleware to ensure that they work as expected, and instead do some integration testing for the external packages, to ensure that they work as expected when integrated into our system.
This will be further described in \autoref{sec:integrationTesting}.
```